From c188872a659cb38028ed61b61d7a51b5c22a2bbb Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:04:22 -0400
Subject: [fs] btrfs: make sure we retry if we couldn't get the page

Message-id: <1407179109-13714-376-git-send-email-sandeen@redhat.com>
Patchwork-id: 87316
O-Subject: [RHEL7.1 PATCH 375/422] Btrfs: make sure we retry if we couldn't get the page
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Filipe Manana <fdmanana@gmail.com>

commit 91405151eb7005949072bee751343390757f8417
Author: Filipe Manana <fdmanana@gmail.com>
Date:   Thu Jun 5 13:22:24 2014 +0100

    Btrfs: make sure we retry if we couldn't get the page

    In inode.c:btrfs_page_exists_in_range(), if we can't get the page
    we need to retry. However we weren't retrying because we weren't
    setting page to NULL, which makes the while loop exit immediately
    and will make us call page_cache_release after exiting the loop
    which is incorrect because our page get didn't succeed. This could
    also make us return true when we shouldn't.

    Signed-off-by: Filipe David Borba Manana <fdmanana@gmail.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index d316b64..a6be9cc 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -6772,8 +6772,10 @@ bool btrfs_page_exists_in_range(struct inode *inode, loff_t start, loff_t end)
    break; /* TODO: Is this relevant for this use case? */
   }
 
-  if (!page_cache_get_speculative(page))
+  if (!page_cache_get_speculative(page)) {
+   page = NULL;
    continue;
+  }
 
   /*
    * Has the page moved?
-- 
1.7.1